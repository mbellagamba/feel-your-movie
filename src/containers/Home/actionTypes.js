export const FETCH_GENRES = 'FETCH_GENRES';
export const FETCH_GENRES_SUCCESS = 'FETCH_GENRES_SUCCESS';
