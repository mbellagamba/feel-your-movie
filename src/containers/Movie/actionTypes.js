export const FETCH_MOVIE = 'FETCH_MOVIE';
export const FETCH_MOVIE_SUCCESS = 'FETCH_MOVIE_SUCCESS';
