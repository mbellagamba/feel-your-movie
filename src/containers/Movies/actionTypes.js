export const DISCOVER_MOVIES = 'DISCOVER_MOVIES';
export const DISCOVER_MOVIES_SUCCESS = 'DISCOVER_MOVIES_SUCCESS';
