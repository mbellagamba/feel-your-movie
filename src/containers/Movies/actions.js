import {
  DISCOVER_MOVIES,
  DISCOVER_MOVIES_SUCCESS,
} from './actionTypes';

export const discoverMovies = () => ({
  type: DISCOVER_MOVIES,
});

export const discoverMoviesSuccess = (movies) => ({
  type: DISCOVER_MOVIES_SUCCESS,
  movies,
});
