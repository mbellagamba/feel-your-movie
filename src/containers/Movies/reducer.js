export const DISCOVER_MOVIES = 'DISCOVER_MOVIES';
export const DISCOVER_MOVIES_SUCCESS = 'DISCOVER_MOVIES_SUCCESS';

export const discoverMovies = () => ({
  type: DISCOVER_MOVIES,
});

export const discoverMoviesSuccess = (movies) => ({
  type: DISCOVER_MOVIES_SUCCESS,
  movies,
});

const reducer = (state = {}, action) => {
  const { type } = action;
  switch (type) {
    case DISCOVER_MOVIES_SUCCESS:
      return {
        ...state,
        movies: action.movies,
      };
    default:
      return state;
  }
};

export default reducer;
