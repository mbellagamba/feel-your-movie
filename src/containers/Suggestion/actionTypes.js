export const SET_SUGGESTION_PARAMS = 'SET_SUGGESTION_PARAMS';
export const CLEAR_SUGGESTION_PARAMS = 'CLEAR_SUGGESTION_PARAMS';
