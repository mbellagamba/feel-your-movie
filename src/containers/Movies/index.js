import React, { useEffect } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { useQueryParams } from '../../hooks';
import * as actions from './actions';
import MovieCard from '../../components/MovieCard';

const Grid = styled.div`
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
`;

const Movies = (props) => {
  const {
    discoverMovies,
    movies,
  } = props;
  const query = useQueryParams();
  const genre = query.has('genre') ? query.get('genre') : undefined;

  useEffect(() => {
    discoverMovies(genre);
  }, [discoverMovies, genre]);

  return (
    <Grid>
      {movies && movies.map((movie) => (
        <Link key={movie.id} to={`/movies/${movie.id}`}>
          <MovieCard
            title={movie.title}
            cover={movie.backdrop_path}
            releaseDate={movie.release_date}
            voteAverage={movie.vote_average}
            voteCount={movie.vote_count}
          />
        </Link>
      ))}
    </Grid>
  );
};

Movies.propTypes = {
  discoverMovies: PropTypes.func.isRequired,
  movies: PropTypes.arrayOf(Object).isRequired,
};

const mapStateToProps = (state) => ({
  ...state.movies,
});

const mapDispatchToProps = {
  ...actions,
};

export default connect(mapStateToProps, mapDispatchToProps)(Movies);
