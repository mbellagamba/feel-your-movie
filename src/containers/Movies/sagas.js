import {
  put,
  takeLatest,
} from 'redux-saga/effects';
import { DISCOVER_MOVIES } from './actionTypes';
import { discoverMoviesSuccess } from './actions';
import * as api from '../../api';

export function* discoverMovies(action) {
  const url = api.discoverMovies(action.genre, action.page);
  const response = yield fetch(url)
    .then((res) => res.json())
    .catch((err) => err);
  yield put(discoverMoviesSuccess(response.results));
}

export function* watchDiscoverMovies() {
  yield takeLatest(DISCOVER_MOVIES, discoverMovies);
}
