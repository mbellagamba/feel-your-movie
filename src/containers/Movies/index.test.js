import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import Movies from './index';
import { DISCOVER_MOVIES } from './actionTypes';

function createMovie(id) {
  return {
    id,
    title: `movie ${id}`,
    vote_average: 0,
    vote_count: 0,
    release_date: '2019',
  };
}

jest.mock('react-router-dom', () => ({
  Link: () => <div />,
  useLocation: jest.fn(() => 'fake-route'),
}));

test('should match snapshot without movies', () => {
  const mockStore = configureStore();
  const store = mockStore({
    movies: {
      movies: [],
      loading: false,
    },
  });
  const { asFragment } = render((
    <Provider store={store}>
      <Movies />
    </Provider>
  ));
  expect(asFragment()).toMatchSnapshot();
});

test('should match snapshot', () => {
  const length = 3;
  const movies = [...Array(length).keys()].map(createMovie);
  const mockStore = configureStore();
  const store = mockStore({
    movies: {
      movies,
      loading: false,
    },
  });
  const { asFragment } = render((
    <Provider store={store}>
      <Movies />
    </Provider>
  ));
  expect(asFragment()).toMatchSnapshot();
});

test('should show movies list', () => {
  const length = 3;
  const movies = [...Array(length).keys()].map(createMovie);
  const mockStore = configureStore();
  const store = mockStore({
    movies: {
      movies,
      loading: false,
    },
  });
  const { getByTestId } = render((
    <Provider store={store}>
      <Movies />
    </Provider>
  ));
  expect(getByTestId('movies-grid').children.length).toEqual(length);
});

test('should not show any movie if list is empty', () => {
  const mockStore = configureStore();
  const store = mockStore({
    movies: {
      movies: [],
      loading: false,
    },
  });
  const { getByText } = render((
    <Provider store={store}>
      <Movies />
    </Provider>
  ));
  expect(getByText('No movie found')).toBeInTheDocument();
});

test('should dispatch DISCOVER_MOVIES action on mount', () => {
  const mockStore = configureStore();
  const store = mockStore({
    movies: {
      movies: [],
      loading: false,
    },
  });
  render((
    <Provider store={store}>
      <Movies />
    </Provider>
  ));
  expect(store.getActions()[0].type).toEqual(DISCOVER_MOVIES);
});

test('should show load component if it is loading', () => {
  const mockStore = configureStore();
  const store = mockStore({
    movies: {
      movies: [],
      loading: true,
    },
  });
  const { getByText } = render((
    <Provider store={store}>
      <Movies />
    </Provider>
  ));
  expect(getByText('Loading')).toBeInTheDocument();
});

test('should load more items on button click', () => {
  const mockStore = configureStore();
  const store = mockStore({
    movies: {
      movies: [createMovie(2)],
      loading: false,
    },
  });
  const { getByText } = render((
    <Provider store={store}>
      <Movies />
    </Provider>
  ));
  fireEvent.click(getByText('Load more'));
  // The action DISCOVER_MOVIES should be dispatched 2 times
  // On mount and on load more click
  const actions = store.getActions()
    .filter((a) => a.type === DISCOVER_MOVIES);
  expect(actions).toHaveLength(2);
});
