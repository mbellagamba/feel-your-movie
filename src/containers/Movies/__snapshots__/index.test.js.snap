// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  color: white;
  line-height: 1.8rem;
  border-radius: 0.9rem;
  border: 1px solid #3dda84;
  margin: 16px 0;
  padding: 0 16px;
}

@media (max-width:425px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
      data-testid="movies-grid"
    >
      <div />
      <div />
      <div />
    </div>
    <button
      class="c2"
      type="button"
    >
      Load more
    </button>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot without movies 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  color: white;
  line-height: 1.8rem;
  border-radius: 0.9rem;
  border: 1px solid #3dda84;
  margin: 16px 0;
  padding: 0 16px;
}

@media (max-width:425px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
      data-testid="movies-grid"
    >
      <span>
        No movie found
      </span>
    </div>
    <button
      class="c2"
      type="button"
    >
      Load more
    </button>
  </div>
</DocumentFragment>
`;
